### DISCUSSION TITLE
Deprecation of Feeds API support for XML and Flat File Listings Feeds on March 2025

### DISCUSSION THREAD

**Discussion by jaylee71** [abuzuhri/Amazon-SP-API-CSharp#720] [General]
Hello

Received below email from MWS Announcements today. Looks like they are getting rid of XML feeds and moving to JSON. Are there any example codes of how to migrate to using JSON from XML?

Thank you

Jay

//------------------------------------------------------------------------------------------

Starting today, we are deprecating XML and flat file listings data (including pricing, inventory, relationships and images) submitted via the Feeds API and will no longer provide new features or categories in the legacy category XML format. If you have an application that uses XML and flat file listings data, you have until March 31, 2025 to migrate to the JSON-based Listings Items APIs or the JSON_LISTINGS_FEED submitted via the Feeds API. This deprecation does not impact any listings feeds submitted by sellers directly through Seller Central.

In 2021, we launched JSON-based Listings APIs and the JSON_LISTINGS_FEED feed type for the Feeds API. These solutions empower developers to access new listings capabilities and are available for all product types in hardlines, softlines, and consumables categories worldwide. With these solutions, developers can build sophisticated workflows to create and manage listings. This change applies to all stores worldwide.

**Feed types that will be sunset on March 31, 2025:**

POST_PRODUCT_DATA
POST_INVENTORY_AVAILABILITY_DATA
POST_PRODUCT_OVERRIDES_DATA
POST_PRODUCT_PRICING_DATA
POST_PRODUCT_IMAGE_DATA
POST_PRODUCT_RELATIONSHIP_DATA
POST_FLAT_FILE_INVLOADER_DATA
POST_FLAT_FILE_LISTINGS_DATA
POST_FLAT_FILE_BOOKLOADER_DATA
POST_FLAT_FILE_CONVERGENCE_LISTINGS_DATA
POST_FLAT_FILE_LISTINGS_DATA
POST_FLAT_FILE_PRICEANDQUANTITYONLY_UPDATE_DATA
POST_UIEE_BOOKLOADER_DATA

**What action is required?**
If you have an application that uses any of the feed types listed above, you have until March 31, 2025 to migrate to the JSON-based Listings Items APIs or the JSON_LISTINGS_FEED submitted via the Feeds API.

You may notice an increase in product attribute requirements in JSON Schemas from the Product Type Definitions API when compared to category XSDs. These changes more accurately represent the actual requirements for Selling Partners (including when attributes are required conditionally based on other listing data provided). Developers are empowered to accurately meet listing requirements before submitting to Amazon, reducing errors previously received after submission.

Feeds API usage for the feed types listed above will enter a grace period early next year, at which time throughput may be degraded.

**Planned new features**
Support for Amazon Business (B2B) pricing, media categories, and digital categories coming later this year. In addition to the current features of the JSON-based Listings APIs and the JSON_LISTINGS_FEED feed type, we will soon launch new features that further empower developers to enhance their listings workflows.

**Comment by wszwc** [comment]
Hello, I was actually working on json publishing products before this news was announced. What I can tell you is that they only provided the most basic publishing structure.
It is recommended that you wait until it is close to decommissioning before updating the structure. By then everyone will switch to json publishing, and the ecosystem will be much better.
Although I have not encountered any problems that cannot be solved, it is very difficult to implement the entire login structure published by json. The following are the problems:
1. Their new architecture is very unfavorable for creating HTLM forms and requires recursion or flattening.
2. Their new architecture includes logical verification, which can obtain constraints between fields before publishing. For example, some fields are not filled in but must be filled in. This sounds good, but the shortcomings are actually huge. After practice, I found that only required fields can be obtained, but it is still difficult to obtain disabled fields, or the linkage between attributes, which makes this The verification became incomplete, but I couldn't bear to abandon it.
3. Their data structure is more redundant. Even if they set each attribute into an array that can fill in multiple values, there are still other_image1, 2, 3, 4, 5, etc. Moreover, the constraints between some enumerations are missing. For example, age is selected as the size unit, but the size value will still have values in other size units.
4. In the new listing mechanism, you can use the interface to get errors synchronously to list products, but you can only list one product at a time, which is fatal for product distribution. So if you want to distribute goods, you still need to use the json submission of the feed. This is the official solution.

**Reply by wszwc** [reply]
By the way, the current SubmitFeed method of this SDK is wrong for json publishing because it does not set the encoding format of http request upload to utf-8 when uploading json data.
But you can avoid this problem by following the official standard steps for submitting feeds and doing it yourself.
The sdk provides corresponding methods.@abuzuhri

**Comment by optitsolutions** [comment]
Sehr geehrte Damen und Herren,

bis inklusive 03.04.2024 bin ich nicht im B++ro und lese meine Mails nur unregelm+ñ+ig.

Bitte wenden Sie sich in dringenden F+ñllen an das B++ro unter ***@***.***

Ihre Mail wird nicht automatisch weitergeleitet.




Mit freundlichen Gr+++en

Thorsten Altmann






Hello, I was actually working on json publishing products before this news was announced. What I can tell you is that they only provided the most basic publishing structure.
It is recommended that you wait until it is close to decommissioning before updating the structure. By then everyone will switch to json publishing, and the ecosystem will be much better.
Although I have not encountered any problems that cannot be solved, it is very difficult to implement the entire login structure published by json. The following are the problems:
1. Their new architecture is very unfavorable for creating HTLM forms and requires recursion or flattening.
2. Their new architecture includes logical verification, which can obtain constraints between fields before publishing. For example, some fields are not filled in but must be filled in. This sounds good, but the shortcomings are actually huge. After practice, I found that only required fields can be obtained, but it is still difficult to obtain disabled fields, or the linkage between attributes, which makes this The verification became incomplete, but I couldn't bear to abandon it.
3. Their data structure is more redundant. Even if they set each attribute into an array that can fill in multiple values, there are still other_image1, 2, 3, 4, 5, etc. Moreover, the constraints between some enumerations are missing. For example, age is selected as the size unit, but the size value will still have values in other size units.
4. In the new listing mechanism, you can use the interface to get errors synchronously to list products, but you can only list one product at a time, which is fatal for product distribution. So if you want to distribute goods, you still need to use the json submission of the feed. This is the official solution.

-- 
Reply to this email directly or view it on GitHub:
https://github.com/abuzuhri/Amazon-SP-API-CSharp/discussions/720#discussioncomment-8962548

**Comment by optitsolutions** [comment]
Sehr geehrte Damen und Herren,

bis inklusive 03.04.2024 bin ich nicht im B++ro und lese meine Mails nur unregelm+ñ+ig.

Bitte wenden Sie sich in dringenden F+ñllen an das B++ro unter ***@***.***

Ihre Mail wird nicht automatisch weitergeleitet.




Mit freundlichen Gr+++en

Thorsten Altmann






By the way, the current SubmitFeed method of this SDK is wrong for json publishing because it does not set the encoding format of http request upload to utf-8 when uploading json data.
But you can avoid this problem by following the official standard steps for submitting feeds and doing it yourself.
The sdk provides corresponding ***@***.***

-- 
Reply to this email directly or view it on GitHub:
https://github.com/abuzuhri/Amazon-SP-API-CSharp/discussions/720#discussioncomment-8962582

**Comment by ihorkolucky** [comment]
Hi. Currently there is only GetXML()

            ConstructFeedService createDocument = new ConstructFeedService("{SellerID}", "1.02");

            var list = new List<PriceMessage>();
            list.Add(new PriceMessage()
            {
                SKU = "8201031206122...",
                StandardPrice = new StandardPrice()
                {
                    currency = amazonConnection.GetCurrentMarketplace.CurrencyCode.ToString(),
                    Value = (201.0522M).ToString("0.00")
                }
            });
            createDocument.AddPriceMessage(list);

            var xml = createDocument.GetXML();

Will you add GetJSON() method soon?

Thank you!

**Reply by huntmj01** [reply]
@ihorkolucky 

I have opened a [PR](https://github.com/abuzuhri/Amazon-SP-API-CSharp/pull/782) to add a method that can be used to parse the JSON_LISTINGS_FEED result.

My [PR](https://github.com/abuzuhri/Amazon-SP-API-CSharp/pull/782) also includes in the README and in FeedsSample.cs an example of how to submit a price change using the JSON_LISTINGS_FEED.

**Comment by WebthinkingAbi** [comment]
Hi, 

With XML being deprecated for the feed I have begun to move over to using JSON_LISTINGS_FEED and when I am going to check the status in code for this feed I am now using GetFeedDocumentProcessingReportAsync instead of GetFeedDocumentProcessingReport as it is being deprecated but it just hangs waiting for a response and I was wondering if there is a fix for this?

**Comment by Barakros** [comment]
hi
I would also want to know if and how we can use the new json feed?

**Reply by Barakros** [reply]
hi
after playing with it
I was able to update qty and price
here is an example for qty:
feed_path is a path to a json file of JSON_LISTINGS_FEED 
the json need to be stored on disk (did not worked for me a a string, maybe need a stream?)

```
var feedID = await amazon_connection.Feed.SubmitFeedAsync(feed_path, FeedType.JSON_LISTINGS_FEED, new List<string>() { MarketPlace.US.ID }, null, ContentType.JSON);

log.WriteLogLine($"feed id: {feedID}");

//loop until feed is ready
FikaAmazonAPI.AmazonSpApiSDK.Models.Feeds.Feed feedOutput = null;
ProcessingStatusEnum proc_status = ProcessingStatusEnum.INPROGRESS;
while (proc_status == ProcessingStatusEnum.INPROGRESS | proc_status == ProcessingStatusEnum.INQUEUE)
{
    log.WriteLogLine($"Status: {proc_status} Waiting 30 seconds...");
    await Task.Delay(AmazonTask.SECOND * 30);
    feedOutput = await amazon_connection.Feed.GetFeedAsync(feedID);
    proc_status = feedOutput.ProcessingStatus;
}

log.WriteLogLine($"Feed Status: {proc_status}");
```
json example:
{"header":{"sellerId":"SELLER_ID","version":"2.0","issueLocale":"en_US"},"messages":[{"messageId":1,"sku":"SKU","operationType":"PATCH","productType":"PRODUCT","patches":[{"op":"replace","path":"/attributes/fulfillment_availability","value":[{"fulfillment_channel_code":"DEFAULT","quantity":QTY}]}]}]}

I still did not found how to set PRIME or not PRIME...
if someone has an idea...
thx

**Reply by WebthinkingAbi** [reply]
Hi, Have you got a way to do it without storing it on the disk?

**Reply by Barakros** [reply]
no

**Reply by VincentMatthews** [reply]
I was able to send the JSON as a string without storing it as a file. It helped to set the format:
`feedID = amazonConnection.Feed.SubmitFeed(json, FeedType.JSON_LISTINGS_FEED, new List<string>() { defaultMarketplace.ID }, null, ContentType.JSON, ContentFormate.Text);`

The default "ContentFormate" is AutoDetect. Inspecting the source code shows that AutoDetect tries to work out if it is a file with the following:

```
        private bool IsPathRooted(string content)
        {
            if (string.IsNullOrEmpty(content))
                return false;

            if (content.Length > 255 || content.Contains("\n"))
                return false;

            return true;
        }
```

If the JSON string is unformatted and less thn 255 chars then it will be regarded as a file.

**Comment by VincentMatthews** [comment]
I have been working through converting the product feeds to JSON and I have run into these issues so far with the existing API code:

1) Document create is building XML - no ConstructFeedService.GetJSON()
Solution : generate JSON myself and submit as string, as per Barakros above. Specify feed content format as Text.
For info on generating the right JSON, this is a good place to start: [https://github.com/amzn/selling-partner-api-models/issues/2372](url)

2) GetFeedDocumentProcessingReportAsync is trying to deserialise as XML and gives an error. The report format can be either JSON or XML, depending on the format of the feed that gets sent. 
Solution : get the report myself (when JSON) and deserialise to a class modelled on [https://github.com/amzn/selling-partner-api-models/blob/main/schemas/feeds/listings-feed-processing-report-schema-v2.json](url)
e.g.
```
        private FeedProcessingReport_JSON GetJSONFeedReport(FeedDocument feedDocument)
        {
            var reportOutput = feedDocument.Url;

            byte[] compressed = null;
            string output;

            using (var wc = new System.Net.WebClient())
                compressed = wc.DownloadData(new Uri(feedDocument.Url));

            using (var stream = new MemoryStream(compressed))
            {
                if (feedDocument.CompressionAlgorithm.HasValue && (feedDocument.CompressionAlgorithm.Value == FeedDocument.CompressionAlgorithmEnum.GZIP))
                {
                    using (var gzipStream = new System.IO.Compression.GZipStream(stream, System.IO.Compression.CompressionMode.Decompress))
                    {
                        using (var gunzipStream = new MemoryStream())
                        {
                            gzipStream.CopyTo(gunzipStream);
                            output = Encoding.UTF8.GetString(gunzipStream.ToArray());
                        }
                    }
                }
                else
                {
                    output = Encoding.UTF8.GetString(stream.ToArray());
                }
            }

            var jsonReport = JsonConvert.DeserializeObject<FeedProcessingReport_JSON>(output);

            return jsonReport;
        }
```

**Comment by huntmj01** [comment]
I have opened a [PR](https://github.com/abuzuhri/Amazon-SP-API-CSharp/pull/782) to add a method that can be used to parse the JSON_LISTINGS_FEED result.

My [PR](https://github.com/abuzuhri/Amazon-SP-API-CSharp/pull/782)  also includes in the README and in FeedsSample.cs an example of how to submit a price change using the JSON_LISTINGS_FEED.
